### ISSUE TITLE
DeltaTDM as the match balancing engine

### ISSUE THREAD

**Issue by deferred-impact** [deferred-impact/battledudes-issue-tracker#224]
I'm not saying DTDM should be the only gamemode. But what i am saying is, it should be used in the background to decide team balance. Keep track of the kill delta and use that to rebalance teams. It really works, i think it really does.

Factors to consider:
 - team balance >>> equal number of players on each team.  Player skills differ, and it's better to have teams uneven in the number of players, than teams uneven in skill. I hope this is obviously true to the reader
 - team balancing is only really possible when players are joining the game. If nobody is joining, how do you balance the teams? Not true. Don't be afraid to switch players between teams. Pick some player (using their own kill delta - read more below), show them a prompt: "hey, if you switch teams now, you'll get 2 gems". If they accept, switch them.

How to pick which player to move between teams? We need some way to compare player skill, but how?

Well, why not use another kill delta yet again. Take all the kills made by all players in the current game, find the average kill count. Now take the player you want to measure, find the difference between their kill count and the global average kill count. There's their skill rating. Use it to decide which player to swap

Edit: note the above system would require you to set a newly joining player's kill score to the current average kill count. This is only for balancing purposes; otherwise their amount of kills is as you would expect, and starts from 0. This starting discrepancy is needed to address people joining late into the game and being rated way too low because they are behind on kills

This game needs proper team balancing NOW. Unbalanced teams ruin games. Properly balanced teams offer the best and most engaging gameplay experience, but are very rare.

**Comment by Haox09** [comment]
can kdr be used for balance or nah

**Comment by deferred-impact** [comment]
Yes and no. KDR is a rather splendid signal for measuring skill, as long as you know it. Trouble is, you don't always know it.  Some people play without logging in, and some of them are pretty good actually.

Additionally, KDR gives perhaps a bit too much weight to not dying. Someone could have 2 KDR and get 60 kills per game, i have 2 KDR and get 20 kills per game. Yet it would be silly to say our impact on the game is the same.

edit: per game, not per minute
